libs/optimizer/src/lib/optimizer.ts

```diff
@@ -0,0 +1,348 @@
+ * @param track the source track
+ */
+function toIgcFile(track: ScoringTrack): IGCFile {
+  const fixes = track.points.map((point): BRecord => {
+    const timeMilliseconds = point.timeSec * 1000;
+    return {
+      timestamp: timeMilliseconds,
+      time: new Date(timeMilliseconds).toISOString(),
+      latitude: point.lat,
+      longitude: point.lon,
+      valid: true,
+      pressureAltitude: null,
+      gpsAltitude: point.alt,
+      extensions: {},
+      fixAccuracy: null,
+      enl: null,
+    };
+  });
+  // we ignore some properties of the igc-file, as they are not required for the computation
+  // @ts-ignore
+  return {
+    date: new Date(track.startTimeSec * 1000).toISOString(),
+    fixes: fixes,
+  };
+}
+
+function toScoringRules(league: ScoringRules) {
+    return scoringRules.get(league);
+}
+
+type SolverOptions = { maxloop?: number; maxcycle?: number };
+
+function toOptions(options?: OptimizationOptions): SolverOptions {
+  return {
+    maxcycle: options?.maxCycleDurationMs,
+    maxloop: options?.maxNumCycles,
+  };
+}
+
+function toResult(solution: Solution, originalTrack: ScoringTrack, solverTrack: ScoringTrack): OptimizationResult {
+  return {
+    score: solution.score ?? 0,
+    lengthKm: solution.scoreInfo?.distance ?? 0,
+    multiplier: solution.opt.scoring.multiplier,
+    circuit: toCircuitType(solution.opt.scoring.code),
+    closingRadius: getClosingRadius(solution),
+    solutionIndices: getIndices(solution, originalTrack, solverTrack),
+    optimal: solution.optimal || false,
+  };
+}
+
+function getClosingRadius(solution: Solution) {
+  // @ts-ignore
+  const closingDistanceFixed: number | undefined = solution.opt.scoring?.closingDistanceFixed;
+  // @ts-ignore
+  const closingDistanceRelativeRatio: number | undefined = solution.opt.scoring?.closingDistanceRelative;
+  const closingDistanceRelative = solution.scoreInfo?.distance && closingDistanceRelativeRatio ?
+    closingDistanceRelativeRatio * solution.scoreInfo?.distance :
+    undefined;
+  const closingDistance = solution.scoreInfo?.cp?.d;
+  if (closingDistance === undefined) {
+    return undefined;
+  }
+  if (closingDistanceFixed !== undefined && closingDistance < closingDistanceFixed) {
+    return closingDistanceFixed;
+  } else if (closingDistanceRelative !== undefined && closingDistance < closingDistanceRelative) {
+    return closingDistanceRelative;
+  }
+  return undefined;
+}
+
+type CircuitTypeCode = 'od' | 'tri' | 'fai' | 'oar';
+
+function toCircuitType(code: string) {
+  switch (code as CircuitTypeCode) {
+    case 'od':
+      return CircuitType.OpenDistance;
+    case 'fai':
+      return CircuitType.FaiTriangle;
+    case 'oar':
+      return CircuitType.OutAndReturn;
+    case 'tri':
+      return CircuitType.FlatTriangle;
+  }
+}
+
+
+// return indices of solution points
+function getIndices(solution: Solution, originalTrack: ScoringTrack, solverTrack: ScoringTrack) {
```

Comment:
Please add a comment on why this is needed and what/how it does

Reply:
Ok. Comment added